cpp/include/cuml/tsa/arima_common.h

```diff
@@ -0,0 +1,210 @@
+
+  inline bool need_prep() const { return static_cast<bool>(d + D + k); }
+};
+
+/**
+ * Structure to hold the parameters (makes it easier to pass as an argument)
+ * @note: the qualifier const applied to this structure will only guarantee
+ *        that the pointers are not changed, but the user can still modify the
+ *        arrays when using the pointers directly!
+ */
+template <typename DataT>
+struct ARIMAParams {
+  DataT* mu = nullptr;
+  DataT* ar = nullptr;
+  DataT* ma = nullptr;
+  DataT* sar = nullptr;
+  DataT* sma = nullptr;
+  DataT* sigma2 = nullptr;
+
+  /**
+   * Allocate all the parameter device arrays
+   *
+   * @tparam      AllocatorT Type of allocator used
+   * @param[in]   order      ARIMA order
+   * @param[in]   batch_size Batch size
+   * @param[in]   alloc      Allocator
+   * @param[in]   stream     CUDA stream
+   * @param[in]   order      ARIMA hyper-parameters
+   * @param[in]   batch_size Number of time series analyzed
+   * @param[in]   tr         Whether these are the transformed parameters
+   */
+  template <typename AllocatorT>
+  void allocate(const ARIMAOrder& order, int batch_size, AllocatorT& alloc,
+                cudaStream_t stream, bool tr = false) {
+    if (order.k && !tr)
+      mu = (DataT*)alloc->allocate(batch_size * sizeof(DataT), stream);
+    if (order.p)
+      ar =
+        (DataT*)alloc->allocate(order.p * batch_size * sizeof(DataT), stream);
+    if (order.q)
+      ma =
+        (DataT*)alloc->allocate(order.q * batch_size * sizeof(DataT), stream);
+    if (order.P)
+      sar =
+        (DataT*)alloc->allocate(order.P * batch_size * sizeof(DataT), stream);
+    if (order.Q)
+      sma =
+        (DataT*)alloc->allocate(order.Q * batch_size * sizeof(DataT), stream);
+    if (!tr)
+      sigma2 = (DataT*)alloc->allocate(batch_size * sizeof(DataT), stream);
+  }
+
+  /**
+   * Deallocate all the parameter device arrays
+   *
+   * @tparam      AllocatorT Type of allocator used
+   * @param[in]   order      ARIMA order
+   * @param[in]   batch_size Batch size
+   * @param[in]   alloc      Allocator
+   * @param[in]   stream     CUDA stream
+   * @param[in]   tr         Whether these are the transformed parameters
+   */
+  template <typename AllocatorT>
+  void deallocate(const ARIMAOrder& order, int batch_size, AllocatorT& alloc,
+                  cudaStream_t stream, bool tr = false) {
+    if (order.k && !tr)
+      alloc->deallocate(mu, batch_size * sizeof(DataT), stream);
+    if (order.p)
+      alloc->deallocate(ar, order.p * batch_size * sizeof(DataT), stream);
+    if (order.q)
+      alloc->deallocate(ma, order.q * batch_size * sizeof(DataT), stream);
+    if (order.P)
+      alloc->deallocate(sar, order.P * batch_size * sizeof(DataT), stream);
+    if (order.Q)
+      alloc->deallocate(sma, order.Q * batch_size * sizeof(DataT), stream);
+    if (!tr) alloc->deallocate(sigma2, batch_size * sizeof(DataT), stream);
+  }
```

Comment:
Thinking out loud... What do you feel about moving these inside constructor and destructor methods respectively?

Reply:
I don't think I should do it, because most of the time the object is constructed in Python and the allocations are not done with these methods. In the benchmark PR I even added two constructors: a default one and one that takes the pointers and write them in the structure.

Reply:
I could have a constructor that takes the order, batch_size, allocator and stream and does what `allocate` does. But I can't have a destructor that does what `deallocate` does because sometimes the structure doesn't own the pointers. So for consistency I think I should have neither